File: java/lang/src/main/java/otterop/lang/PureIterator.java
package otterop.lang;

import java.util.Iterator;

public class PureIterator<T> implements Iterator<T> {

    private OOPIterator<T> oopIterator;

    private PureIterator(OOPIterator<T> oopIterator) {
        this.oopIterator = oopIterator;
    }

    public static <T> Iterator<T> newIterator(OOPIterator<T> oopIterator) {
        return new PureIterator<T>(oopIterator);
    }

    @Override
    public boolean hasNext() {
        return this.oopIterator.hasNext();
    }

    @Override
    public T next() {
        return this.oopIterator.next();
    }
}

File: dotnet/src/Lang/Otterop/Lang/PureIterator.cs
namespace Otterop.Lang;

public class PureIterator
{
    public static IEnumerator<T> NewIterator<T>(OOPIterator<T> Iterator)
    {
        return new PureIterator<T>(Iterator);
    }
}

public class PureIterator<T> : System.Collections.Generic.IEnumerator<T>, System.Collections.IEnumerator
{
    private OOPIterator<T> Iterator;

    public T _Current = default(T);

    internal PureIterator(OOPIterator<T> Iterator)
    {
        this.Iterator = Iterator;
    }

    public bool MoveNext()
    {
        bool ret = this.Iterator.HasNext();
        if (ret)
        {
            this._Current = this.Iterator.Next();
        }
        return ret;
    }

    public void Reset()
    {
        throw new NotSupportedException();
    }

    T System.Collections.Generic.IEnumerator<T>.Current
    {
        get { return this._Current; }
    }
    object System.Collections.IEnumerator.Current
    {
        get { return this._Current; }
    }

    public void Dispose()
    {

    }

    public static IEnumerator<T> NewIterator(OOPIterator<T> Iterator)
    {
        return new PureIterator<T>(Iterator);
    }
}
